services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/limits/ProjectStorageLimitManager.java

```diff
@@ -0,0 +1,127 @@
+		
+	private static final Comparator<ProjectStorageLocationUsage> LOCATION_COMPARATOR = Comparator.comparing(ProjectStorageLocationUsage::getStorageLocationId);
+	
+	private TransactionalMessenger messenger;
+	
+	private ProjectStorageLimitsDao storageUsageDao;
+	
+	private TableIndexDAO replicationDao;
+	
+	private NodeDAO nodeDao;
+	
+	private Clock clock;
+	
+	private Set<Long> accessedProjects;
+	
+	public ProjectStorageLimitManager(TransactionalMessenger messenger, ProjectStorageLimitsDao storageUsageDao, TableIndexDAO replicationDao, NodeDAO nodeDao, Clock clock) {
+		this.messenger = messenger;
+		this.storageUsageDao = storageUsageDao;
+		this.replicationDao = replicationDao;
+		this.nodeDao = nodeDao;
+		this.clock = clock;
+		this.accessedProjects = ConcurrentHashMap.newKeySet();
+	}
+	
+	public ProjectStorageUsage gerProjectStorageUsage(String projectId) {
+		ValidateArgument.requiredNotBlank(projectId, "The projectId");
+		ValidateArgument.requirement(EntityType.project.equals(nodeDao.getNodeTypeById(projectId)), "The entity with the given id is not a project.");
+		
+		Long projectIdLong = KeyFactory.stringToKey(projectId);
+		
+		Map<String, ProjectStorageLocationUsage> locations = new HashMap<>();
+		
+		// First sets the limits
+		storageUsageDao.getStorageLocationLimits(projectIdLong).forEach(limit -> {
+			locations.put(limit.getStorageLocationId(), new ProjectStorageLocationUsage()
+				.setStorageLocationId(limit.getStorageLocationId())
+				.setMaxAllowedFileBytes(limit.getMaxAllowedFileBytes())
+				.setSumFileBytes(0L)
+				.setIsOverLimit(false)
+			);
+		});
+		
+		// Now updates the usage
+		storageUsageDao.getStorageData(projectIdLong)
+			.map(ProjectStorageData::getStorageLocationData)
+			.orElseGet(() -> Collections.emptyMap())
+			.forEach((storageLocationId, currentUsage) -> {
+				ProjectStorageLocationUsage usage = locations.get(storageLocationId);
+				
+				// No limit defined for this location, sets the usage
+				if (usage == null) {
+					locations.put(storageLocationId, new ProjectStorageLocationUsage()
+						.setStorageLocationId(storageLocationId)
+						.setSumFileBytes(currentUsage)
+						.setIsOverLimit(false)
+						.setMaxAllowedFileBytes(null)
+					);
+				// A limit is defined, sets the usage accordingly
+				} else {
+					usage.setSumFileBytes(currentUsage)
+						.setIsOverLimit(currentUsage > usage.getMaxAllowedFileBytes());
+				}
+			});
+
+		accessedProjects.add(projectIdLong);
+						
+		return new ProjectStorageUsage()
+			.setProjectId(KeyFactory.keyToString(projectIdLong))
+			.setLocations(locations.values().stream().sorted(LOCATION_COMPARATOR).collect(Collectors.toList()));
+	}
+	
+	@WriteTransaction
+	public void refreshProjectStorageData(Long projectId) {
+		if (storageUsageDao.isStorageDataModifiedOnAfter(projectId, clock.now().toInstant().minus(CACHE_UPDATE_FREQUENCY))) {
+			return;
+		}
+		
+		ProjectStorageData data = replicationDao.computeProjectStorageData(projectId);
```

Comment:
Note: there is an additional complication: a user can create a S3StorageLocationSetting that points to the default synapse bucket. This storage location will have a different id than the default storage location so this grouping won't work if we want a limit on our bucket per project.